src/inet/UDPEndPointImplLwIP.cpp

```diff
@@ -66,27 +63,64 @@ static_assert(LWIP_TCPIP_CORE_LOCKING, "CHIP requires config LWIP_TCPIP_CORE_LOC
 namespace chip {
 namespace Inet {
 
-namespace {
-/**
- * @brief
- * RAII locking for LwIP core to simplify management of
- * LOCK_TCPIP_CORE()/UNLOCK_TCPIP_CORE() calls.
- */
-class ScopedLwIPLock
+EndpointQueueFilter * UDPEndPointImplLwIP::sQueueFilter = nullptr;
+
+err_t EndPointStateLwIP::RunOnTCPIPRet(std::function<err_t()> fn)
 {
-public:
-    ScopedLwIPLock() { LOCK_TCPIP_CORE(); }
-    ~ScopedLwIPLock() { UNLOCK_TCPIP_CORE(); }
-};
-} // anonymous namespace
+    assertChipStackLockedByCurrentThread();
+#if LWIP_TCPIP_CORE_LOCKING
+    err_t err;
+    LOCK_TCPIP_CORE();
+    err = fn();
+    UNLOCK_TCPIP_CORE();
+    return err;
+#else
+    // Post a message to the TCPIP task and wait for it to run.
+    static sys_sem_t sTCPIPSem;
+    static bool sTCPIPSemInited = false;
+    if (!sTCPIPSemInited)
+    {
+        err_t err = sys_sem_new(&sTCPIPSem, 0);
+        if (err != ERR_OK)
+        {
+            return err;
+        }
+        sTCPIPSemInited = true;
+    }
 
-EndpointQueueFilter * UDPEndPointImplLwIP::sQueueFilter = nullptr;
+    // These are static to be able to share them with the TCPIP callback.
```

Comment:
```suggestion
    // tcpip_callback takes a C function pointer, so we can't pass a capturing lambda to it.
    // Just store the state the function we pass to it needs in statics.
    // This should be safe, since that function will execute before we return and there is no
    // re-entry into this method.
```